server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/InnerSpaceCompactionExceptionHandler.java

```diff
@@ -0,0 +1,207 @@
+      LOGGER.info(
+          "{} [Compaction][ExceptionHandler] some source files {} is lost",
+          fullStorageGroupName,
+          lostSourceFiles);
+      if (!targetTsFile.getTsFile().exists()) {
+        // some source files are missed, and target file not exists
+        // some data is lost, set the system to read-only
+        LOGGER.warn(
+            "{} [Compaction][ExceptionHandler] target file {} does not exist either, do nothing. Set system to read-only",
+            fullStorageGroupName,
+            targetTsFile);
+        IoTDBDescriptor.getInstance().getConfig().setReadOnly(true);
+        handleSuccess = false;
+      } else {
+        handleSuccess =
+            handleWhenSomeSourceFilesLost(
+                fullStorageGroupName,
+                targetTsFile,
+                selectedTsFileResourceList,
+                tsFileResourceList,
+                lostSourceFiles);
+      }
+    }
+
+    if (!handleSuccess) {
+      LOGGER.error(
+          "{} [Compaction][ExceptionHandler] Failed to handle exception, set allowCompaction to false",
+          fullStorageGroupName);
+      tsFileManager.setAllowCompaction(false);
+    } else {
+      LOGGER.info(
+          "{} [Compaction][ExceptionHandler] Handle exception successfully, delete log file {}",
+          fullStorageGroupName,
+          logFile);
+      try {
+        FileUtils.delete(logFile);
+      } catch (IOException e) {
+        LOGGER.error(
+            "{} [Compaction][ExceptionHandler] Exception occurs while deleting log file {}, set allowCompaction to false",
+            fullStorageGroupName,
+            logFile,
+            e);
+        tsFileManager.setAllowCompaction(false);
+      }
+    }
+  }
+
+  private static boolean checkAllSourceFilesExist(
+      List<TsFileResource> sourceFiles, List<TsFileResource> lostSourceFiles) {
+    boolean allSourceFileExist = true;
+    for (TsFileResource sourceTsFile : sourceFiles) {
+      if (!sourceTsFile.getTsFile().exists()) {
+        allSourceFileExist = false;
+        lostSourceFiles.add(sourceTsFile);
+      }
+    }
+    return allSourceFileExist;
+  }
+
+  private static boolean handleWhenAllSourceFilesExist(
+      String fullStorageGroupName,
+      TsFileResource targetTsFile,
+      List<TsFileResource> selectedTsFileResourceList,
+      TsFileResourceList tsFileResourceList) {
+    // all source file exists, delete the target file
+    LOGGER.info(
+        "{} [Compaction][ExceptionHandler] all source files {} exists, delete target file {}",
+        fullStorageGroupName,
+        selectedTsFileResourceList,
+        targetTsFile);
+    if (!targetTsFile.remove()) {
+      // failed to remove target tsfile
+      // system should not carry out the subsequent compaction in case of data redundant
+      LOGGER.warn(
+          "{} [Compaction][ExceptionHandler] failed to remove target file {}",
+          fullStorageGroupName,
+          targetTsFile);
+      return false;
+    }
```

Comment:
.mods file of the targe file should also be deleted here.

Reply:
resolve